app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js

```diff
@@ -234,91 +195,298 @@ async function addEthereumChainHandler(
-    return end(error);
+    // For the purposes of this method, it does not matter if the user
+    // declines to switch the selected network. However, other errors indicate
+    // that something is wrong.
+    if (error.code !== errorCodes.provider.userRejectedRequest) {
+      return end(error);
+    }
+  } finally {
+    if (approvalFlowId) {
+      endApprovalFlow({ id: approvalFlowId });
+    }
   }
+  return end();
+};
 
+const switchChain = async (
+  res,
+  end,
+  origin,
+  chainId,
+  requestData,
+  networkConfigurationId,
+  approvalFlowId,
+  {
+    getChainPermissionsFeatureFlag,
+    setActiveNetwork,
+    endApprovalFlow,
+    requestUserApproval,
+    getCaveat,
+    requestSwitchNetworkPermission,
+  },
+) => {
+  // If this network is already added but is not the currently selected network for the given origin
   // Ask the user to switch the network
-  try {
-    await requestUserApproval({
+  if (getChainPermissionsFeatureFlag()) {
+    await switchChainWithPermissions(
+      res,
+      end,
+      networkConfigurationId,
+      approvalFlowId,
+      chainId,
+      {
+        getCaveat,
+        requestSwitchNetworkPermission,
+        setActiveNetwork,
+        endApprovalFlow,
+      },
+    );
+  } else {
+    // TODO remove once CHAIN_PERMISSIONS feature flag is gone
+    await switchChainWithoutPermissions(
+      res,
+      end,
+      requestData,
+      networkConfigurationId,
+      approvalFlowId,
       origin,
-      type: ApprovalType.SwitchEthereumChain,
-      requestData: {
-        toNetworkConfiguration: {
+      { setActiveNetwork, endApprovalFlow, requestUserApproval },
+    );
+  }
+  return end();
+};
+
+async function addEthereumChainHandler(
+  req,
+  res,
+  _next,
+  end,
+  {
+    upsertNetworkConfiguration,
+    getCurrentRpcUrl,
+    findNetworkConfigurationBy,
+    setActiveNetwork,
+    requestUserApproval,
+    startApprovalFlow,
+    endApprovalFlow,
+    getCurrentChainIdForDomain,
+    getCaveat,
+    requestSwitchNetworkPermission,
+    getChainPermissionsFeatureFlag,
+  },
+) {
+  const validParams = validateParams(req.params[0], end);
+  if (!validParams) {
+    return end();
+  }
+
+  const {
+    chainId,
+    chainName,
+    firstValidBlockExplorerUrl,
+    firstValidRPCUrl,
+    ticker,
+  } = validParams;
+  const { origin } = req;
+
+  const currentChainIdForDomain = getCurrentChainIdForDomain(origin);
+  const currentNetworkConfiguration = findExistingNetwork(
+    currentChainIdForDomain,
+    findNetworkConfigurationBy,
+  );
+
+  // TODO should this be searching by rpcUrl instead of chainId?
+  // currently we aren't preventing adding the same rpcUrl multiple times
+  // since this just finds the first network with the same chainId
+  // the actual add is prevented downstream in upsertNetworkConfiguration
+  // but we should consider changing this to search by rpcUrl
+  const existingNetwork = findExistingNetwork(
+    chainId,
+    findNetworkConfigurationBy,
+  );
```

Comment:
DRY: Notice the similarities between these lines and the top of the `switchEthereumChain` implementation.

Reply:
Did some general DRYing here: https://github.com/MetaMask/metamask-extension/pull/24415/commits/9d07ed684b9484b5f42a3baea873a39aa413be7e